### PR TITLE
Remove non fatal errors from SPIRE client in operator

### LINKED ISSUE #27842: Operator emits lots of `level=error msg="Failed to watch the Workload API"` log messages
I've observed in our CI that the operator logs are repeating the following messages:

```
291:2023-08-29T09:50:40.608250948Z level=error msg="Failed to watch the Workload API: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/spire/sockets/agent/agent.sock: connect: no such file or directory\"" subsys=spire-client
336:2023-08-29T09:50:41.609488564Z level=error msg="Failed to watch the Workload API: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/spire/sockets/agent/agent.sock: connect: no such file or directory\"" subsys=spire-client
339:2023-08-29T09:50:43.612937910Z level=error msg="Failed to watch the Workload API: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/spire/sockets/agent/agent.sock: connect: no such file or directory\"" subsys=spire-client
342:2023-08-29T09:50:46.613703554Z level=error msg="Failed to watch the Workload API: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/spire/sockets/agent/agent.sock: connect: no such file or directory\"" subsys=spire-client
400:2023-08-29T09:50:50.618140619Z level=error msg="Failed to watch the Workload API: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/spire/sockets/agent/agent.sock: connect: no such file or directory\"" subsys=spire-client
408:2023-08-29T09:50:55.624617307Z level=error msg="Failed to watch the Workload API: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/spire/sockets/agent/agent.sock: connect: no such file or directory\"" subsys=spire-client
414:2023-08-29T09:51:01.630419592Z level=error msg="Failed to watch the Workload API: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/spire/sockets/agent/agent.sock: connect: no such file or directory\"" subsys=spire-client
418:2023-08-29T09:51:08.632722409Z level=error msg="Failed to watch the Workload API: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/spire/sockets/agent/agent.sock: connect: no such file or directory\"" subsys=spire-client
424:2023-08-29T09:51:10.811551775Z level=error msg="Failed to watch the Workload API: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/spire/sockets/agent/agent.sock: connect: no such file or directory\"" subsys=spire-client
429:2023-08-29T09:51:11.822885974Z level=error msg="Failed to watch the Workload API: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/spire/sockets/agent/agent.sock: connect: no such file or directory\"" subsys=spire-client
442:2023-08-29T09:51:13.818073617Z level=error msg="Failed to watch the Workload API: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/spire/sockets/agent/agent.sock: connect: no such file or directory\"" subsys=spire-client
446:2023-08-29T09:51:16.819469143Z level=error msg="Failed to watch the Workload API: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/spire/sockets/agent/agent.sock: connect: no such file or directory\"" subsys=spire-client
461:2023-08-29T09:51:20.824275090Z level=error msg="Failed to watch the Workload API: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/spire/sockets/agent/agent.sock: connect: no such file or directory\"" subsys=spire-client
469:2023-08-29T09:51:25.823068085Z level=error msg="Failed to watch the Workload API: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/spire/sockets/agent/agent.sock: connect: no such file or directory\"" subsys=spire-client
475:2023-08-29T09:51:31.823914725Z level=error msg="Failed to watch the Workload API: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/spire/sockets/agent/agent.sock: connect: no such file or directory\"" subsys=spire-client
481:2023-08-29T09:51:38.824253607Z level=error msg="Failed to watch the Workload API: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial unix /run/spire/sockets/agent/agent.sock: connect: no such file or directory\"" subsys=spire-client
```

Those messages are benign, i.e. the operator eventually succeeds in connecting to the workload API.

The `error` level should not be used unless user action is required (many users even have alerts for `level=error` log messages, and so this causes false positives). `level=error` and `level=warning` for benign events is also problematic for tooling which attempts to find issues in the logs.

Some proposals on how to improve this:
- Only start emitting the log message if the connection is not established after a certain time, e.g. one minute (or whatever time we consider reasonable). Examples of this can be found [here](https://github.com/cilium/cilium/blob/c66da33c6111b35a995b649f46c059701bc27167/daemon/cmd/hubble.go#L285-L294) or [here](https://github.com/cilium/cilium/blob/69d7baef77a6079e3061b2f184d077e4972a92e5/pkg/ipam/multipool.go#L299C7-L303). We can still print the connection attempts with `debug` level.
- For `cilium-agent`, we already have a [check](https://github.com/cilium/cilium-cli/blob/d5971d48b2603ff8cd4bcfe8ad1bba37e8e5357e/connectivity/tests/errors.go#L18-L21) that forbids any `level=error` log messages in CI. We should do the same for the operator to avoid regressions.

### PR DESCRIPTION
**Pull request by meyskens**
This removed level=error errors from the spire-client system in the operator. As it is expected to fail on initial startup of SPIRE to have connection issues Cilium will retry.

Some errors also came from the SPIFFE library itself which is why a wrapper to log them as warnings instead was added.

Fixes: https://github.com/cilium/cilium/issues/27842

```release-note
Remove non fatal errors from SPIRE client in the operator
```

### CODE REVIEW

File: operator/auth/spire/log_wrapper.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+// Warnf logs a warning message
+func (l *spiffeLogWrapper) Warnf(format string, args ...interface{}) {
+	l.log.Warnf(format, args...)
+}
+
+// Errorf logs an error message downgraded to a warning as in our case
+// a connection error on startups is expected on initial start of the oprator
+// while the SPIRE server is still starting up. Any error level messages will also
```
Comment (mhofstetter) [inline operator/auth/spire/log_wrapper.go]: nit: 

> Any error level messages

Isn't it the error itself (and not the error level message?) that gets catched by the operators error handler and then written as error level message to the log?

**Review by mhofstetter — APPROVED** [review]
Nice @meyskens 

Only one non-blocking nit in a comment.

File: operator/auth/spire/client.go [hunk 2]
```diff
@@ -135,7 +135,7 @@ func (c *Client) onStart(_ hive.HookContext) error {
 				c.entry = entryv1.NewEntryClient(conn)
 				break
 			}
-			c.log.WithError(err).Errorf("Unable to connect to SPIRE server, attempt %d", attempts+1)
+			c.log.WithError(err).Warnf("Unable to connect to SPIRE server, attempt %d", attempts+1)
```
Comment (pippolo84) [inline operator/auth/spire/client.go:138]: While a warning is surely more appropriate than an error, reading the linked issue, it seems that we could have similar downsides with a warning too. Why not setting something like a reasonable threshold (either in terms of time or retries count) after which starts logging the warning?

File: operator/auth/spire/log_wrapper.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+// be caught by the operator's error handler to be logged as such.
+func (l *spiffeLogWrapper) Errorf(format string, args ...interface{}) {
+	l.log.Warnf(format, args...)
```
Comment (pippolo84) [inline operator/auth/spire/log_wrapper.go:43]: Is it possible that by doing so, we are gonna end up lowering the errors level too aggressively? What about errors that are not related to the expected ones during initialization?

**Review by pippolo84 — COMMENTED** [review]
Thanks for your PR!
I've left a couple of questions inline.

File: operator/auth/spire/log_wrapper.go (hunk 3 above)
Comment (meyskens) [reply operator/auth/spire/log_wrapper.go:43]: We also catch any errors in our interaction with their API where we can surface them again as errors if needed.

File: operator/auth/spire/client.go (hunk 2 above)
Comment (meyskens) [reply operator/auth/spire/client.go:138]: I have been thinking about this but finding a certain time to assume a bootstrap to be ready is hard to define as it depends on many factors. It being not connected is already surfaced in cilium status. 
Suggestions on this welcome, i like the idea but afraid to put a number on it

Comment (pippolo84) [reply operator/auth/spire/client.go:138]: Yep, I see your point and TBH I don't feel comfortable suggesting a number, either. A bit too verbose still sounds better than being too silent, so I'd say let's go with this and see if it'll be possible to define a reasonable threshold later, if needed. :+1:

**Review by pippolo84 — APPROVED** [review]
LGTM, thanks! :100:

**Comment by giorio94** [comment]
:wave: Hey, would it be possible to get this PR in, as it is currently one of the blockers for https://github.com/cilium/cilium-cli/pull/2184? Reviews are already there, and I guess that a rebase should make CI happier :pray:

**Comment by dylandreimerink** [comment]
@meyskens can all open conversations be considered resolved?

**Comment by pippolo84** [comment]
> @meyskens can all open conversations be considered resolved?

Regarding my comments everything is fine, I marked them as resolved. :+1:

**Pull request merged**